packages/foundry/contracts/oracles/crab/CrabOracle.sol

```diff
@@ -0,0 +1,151 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity >=0.8.13;
+
+import "@yield-protocol/utils-v2/contracts/access/AccessControl.sol";
+import "@yield-protocol/utils-v2/contracts/cast/CastBytes32Bytes6.sol";
+import "@yield-protocol/utils-v2/contracts/token/IERC20Metadata.sol";
+import "../../interfaces/IOracle.sol";
+
+interface ICrabStrategy {
+    function totalSupply() external view returns (uint256);
+
+    /**
+     * @notice get the vault composition of the strategy
+     * @return operator
+     * @return nft collateral id
+     * @return collateral amount
+     * @return short amount
+     */
+    function getVaultDetails()
+        external
+        view
+        returns (
+            address,
+            uint256,
+            uint256,
+            uint256
+        );
+}
+
+/**
+ * @title CrabOracle
+ * @notice Oracle to fetch Crab-ETH exchange amounts
+ */
+contract CrabOracle is IOracle, AccessControl {
+    using CastBytes32Bytes6 for bytes32;
+    ICrabStrategy crabStrategy;
+    IOracle uniswapV3Oracle;
+    bytes6 public ethId;
+    bytes6 public crabId;
+    bytes6 public oSQTHId;
+
+    event SourceSet(
+        bytes6 crab_,
+        bytes6 oSQTH_,
+        bytes6 ethId_,
+        ICrabStrategy indexed crabStrategy_,
+        IOracle indexed uniswapV3Oracle_
+    );
+
+    /**
+     * @notice Set crabstrategy & uniswap source
+     */
+    function setSource(
```

Comment:
You can skip this function and set the source on the constructor, which also means that you don't need to inherit AccessControl anymore. Doing that you can also set all state variables to `immutable`.